include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp

```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
-        activeCommA = viewingCommA;
-        activeCommB = viewingCommB;
-        if(!mpi::Congruent(viewingCommA, viewingCommB))
-            LogicError("Viewing communicators were not congruent");
-    }
-    else if(viewingCommSizeA == commSizeB)
-    {
-        usingViewingA = true;
-        usingViewingB = false;
-        activeCommA = viewingCommA;
-        activeCommB = commB;
-        if(!mpi::Congruent(viewingCommA, commB))
-            LogicError("Communicators were not congruent");
-    }
-    else if(commSizeA == viewingCommSizeB)
-    {
-        usingViewingA = false;
-        usingViewingB = true;
-        activeCommA = commA;
-        activeCommB = viewingCommB;
-        if(!mpi::Congruent(commA, viewingCommB))
-            LogicError("Communicators were not congruent");
+
+    Int rowStrideA = A.RowStride();
+    const Int sizeA = A.Grid().VCSize();
+    const Int rowGCD = GCD(rowStrideB, rowStrideA);
+    const Int rowLCM = rowStrideB*rowStrideA / rowGCD;
+
+    // Parent Subgrid Size: 4 Child Subgrid Size: 3
+    // Parent 0 1 2 3 0 1 2 3 0 1 2 3 
+    // Child  0 1 2 0 1 2 0 1 2 0 1 2
+
+    std::vector<bool> require_data(sizeB,false);
+    std::vector<int> index_to_put(sizeB,-1);
+    std::vector<int> index_from(sizeB,-1);
+    Int temp_require_data = Mod(posInGrid, sizeB) ;
+    double total_iter = posInGrid;
+
+    for(Int i=0; i< int(rowLCM/sizeA); ++i)
+    {
+
+        require_data[temp_require_data] = true;
+        index_to_put[temp_require_data] = i;
+        index_from[temp_require_data] = int(std::floor(total_iter/sizeB));
+
+        total_iter += sizeA;
+        temp_require_data = Mod(temp_require_data+sizeA, sizeB );
     }
-    else
+
+    SyncInfo<D1> syncInfoA = SyncInfoFromMatrix(A.LockedMatrix());
+    SyncInfo<D2> syncInfoB = SyncInfoFromMatrix(B->LockedMatrix());
+    SyncInfo<D1> syncGeneral = SyncInfo<D1>();
+
+    //
+
+    const bool inAGrid = A.Participating();
+    const bool inBGrid = indexB >=0 ? true:false;
+
+    const Int maxSendSize = mLocB * nLocB;
+
+    simple_buffer<T,D1> send_buf(maxSendSize , syncInfoB);
+    simple_buffer<T,D2> recv_buf(maxSendSize , syncInfoA);
+
+    T* sendBuf = send_buf.data();
+    T* recvBuf = recv_buf.data();
+
+    for(Int localRankB = 0; localRankB < sizeB; localRankB++)
+    {
+        if(myLocalRankB==localRankB && inBGrid)
+        {
+            copy::util::InterleaveMatrix(
+                mLocB, nLocB,
+                B->LockedBuffer(0,0),
+                1, B->LDim(),
+                sendBuf, 1, mLocB, syncInfoB);
+
+        }
+        else
+        {
+            T val = 0;
+            hydrogen::details::setBufferToValue(sendBuf, maxSendSize, val,syncGeneral);
+        }
+
+        Synchronize(syncGeneral);
+        mpi::AllReduce( sendBuf, recvBuf, maxSendSize, mpi::SUM, viewingCommA,syncInfoB);
+
+
+        if(inAGrid)
+        {
+            if(require_data[localRankB])
+            {
+                int sendWidth = int(n / rowLCM);
```

Comment:
I don't think we're properly handling the case where `n` doesn't evenly divide `rowLCM`:
```suggestion
                int sendWidth = (n + rowLCM - 1) / rowLCM;
                if (localRankB > n % rowLCM)
                {
                    --sendWidth;
                }
```